File: java/src/main/java/com/quant/localvol/Tridiag.java
package com.quant.localvol;

/**
 * Native Thomas tridiagonal solver.
 *
 * <p>Solves {@code A x = d} where {@code A} is tridiagonal with sub-diagonal
 * {@code a} (length {@code n-1}), diagonal {@code b} (length {@code n}) and
 * super-diagonal {@code c} (length {@code n-1}). The Thomas algorithm is
 * Gaussian elimination without pivoting in O(n): it is stable for the
 * diagonally-dominant systems produced by the Crank-Nicolson discretisation
 * (the diagonal {@code 1 + theta dt (2a/h^2 + r)} dominates the off-diagonals
 * whenever the scheme's local Peclet condition {@code |mu| h <= 2 a} holds,
 * which the PDE grids used here satisfy by construction).
 *
 * <p>No library banded solver is used anywhere in the pricing path — this is
 * the single linear-algebra kernel, ported line for line from the Python
 * reference.
 */
public final class Tridiag {

    private static final double PIVOT_TOL = 1e-300;

    private Tridiag() {
    }

    /**
     * Solve a tridiagonal system by the Thomas algorithm.
     *
     * @param sub  sub-diagonal (length {@code n-1}); {@code sub[i]} multiplies
     *             {@code x[i]} in row {@code i+1}
     * @param diag main diagonal (length {@code n})
     * @param sup  super-diagonal (length {@code n-1}); {@code sup[i]} multiplies
     *             {@code x[i+1]} in row {@code i}
     * @param rhs  right-hand side (length {@code n})
     * @return solution vector {@code x} (length {@code n}), freshly allocated
     * @throws IllegalArgumentException on inconsistent lengths, empty system,
     *         non-finite input, or a (near-)zero pivot during elimination
     */
    public static double[] thomasSolve(double[] sub, double[] diag, double[] sup, double[] rhs) {
        if (sub == null || diag == null || sup == null || rhs == null) {
            throw new IllegalArgumentException("thomasSolve: null input array");
        }
        int n = diag.length;
        if (n == 0) {
            throw new IllegalArgumentException("thomasSolve: empty system");
        }
        int off = Math.max(n - 1, 0);
        if (sub.length != off || sup.length != off || rhs.length != n) {
            throw new IllegalArgumentException(String.format(
                    "thomasSolve: inconsistent shapes sub=%d, diag=%d, sup=%d, rhs=%d",
                    sub.length, diag.length, sup.length, rhs.length));
        }
        requireFinite(sub);
        requireFinite(diag);
        requireFinite(sup);
        requireFinite(rhs);

        // Forward elimination: cp/dp hold the modified super-diagonal and rhs.
        double[] cp = new double[n];
        double[] dp = new double[n];
        double piv = diag[0];
        if (Math.abs(piv) < PIVOT_TOL) {
            throw new IllegalArgumentException("thomasSolve: zero pivot at row 0");
        }
        cp[0] = n > 1 ? sup[0] / piv : 0.0;
        dp[0] = rhs[0] / piv;
        for (int i = 1; i < n; i++) {
            piv = diag[i] - sub[i - 1] * cp[i - 1];
            if (Math.abs(piv) < PIVOT_TOL) {
                throw new IllegalArgumentException("thomasSolve: zero pivot at row " + i);
            }
            cp[i] = i < n - 1 ? sup[i] / piv : 0.0;
            dp[i] = (rhs[i] - sub[i - 1] * dp[i - 1]) / piv;
        }

        // Back substitution.
        double[] x = new double[n];
        x[n - 1] = dp[n - 1];
        for (int i = n - 2; i >= 0; i--) {
            x[i] = dp[i] - cp[i] * x[i + 1];
        }
        return x;
    }

    private static void requireFinite(double[] a) {
        for (double v : a) {
            if (!Double.isFinite(v)) {
                throw new IllegalArgumentException("thomasSolve: non-finite input");
            }
        }
    }
}

File: python/src/localvol/tridiag.py
"""Native Thomas tridiagonal solver.

Solves ``A x = d`` where ``A`` is tridiagonal with sub-diagonal ``a`` (length
``n-1``), diagonal ``b`` (length ``n``) and super-diagonal ``c`` (length
``n-1``).  The Thomas algorithm is Gaussian elimination without pivoting in
O(n): it is stable for the diagonally-dominant systems produced by the
Crank-Nicolson discretisation (the diagonal ``1 + theta dt (2a/h^2 + r)``
dominates the off-diagonals for any dt, h > 0 whenever both off-diagonals
are non-negative).  Central differencing gives that only under the mesh
Peclet condition ``|mu| h <= 2 a``, which a flat vol on the default grid
satisfies but a floored local vol with a few percent of carry does not;
the PDE therefore switches those nodes to upwind differencing (see
:mod:`localvol.pde`), which restores non-negative off-diagonals and hence
the diagonal dominance this kernel relies on.

No SciPy banded solver is used anywhere in the pricing path — this module is
the single linear-algebra kernel, so the C++/Rust/Java ports reproduce it
line for line.
"""

from __future__ import annotations

import numpy as np
from numpy.typing import NDArray

__all__ = ["thomas_solve"]

_PIVOT_TOL = 1e-300


def thomas_solve(
    sub: NDArray[np.float64],
    diag: NDArray[np.float64],
    sup: NDArray[np.float64],
    rhs: NDArray[np.float64],
) -> NDArray[np.float64]:
    """Solve a tridiagonal system by the Thomas algorithm.

    Args:
        sub:  sub-diagonal ``a_1..a_{n-1}`` (length ``n-1``); ``sub[i]``
              multiplies ``x[i]`` in row ``i+1``.
        diag: main diagonal ``b_0..b_{n-1}`` (length ``n``).
        sup:  super-diagonal ``c_0..c_{n-2}`` (length ``n-1``); ``sup[i]``
              multiplies ``x[i+1]`` in row ``i``.
        rhs:  right-hand side (length ``n``).

    Returns:
        Solution vector ``x`` (length ``n``), a fresh float64 array.

    Raises:
        ValueError: on inconsistent lengths, empty system, non-finite input,
            or a (near-)zero pivot encountered during elimination.
    """
    a = np.asarray(sub, dtype=np.float64)
    b = np.asarray(diag, dtype=np.float64)
    c = np.asarray(sup, dtype=np.float64)
    d = np.asarray(rhs, dtype=np.float64)
    n = b.shape[0]
    if n == 0:
        raise ValueError("thomas_solve: empty system")
    if a.shape != (max(n - 1, 0),) or c.shape != (max(n - 1, 0),) or d.shape != (n,):
        raise ValueError(
            f"thomas_solve: inconsistent shapes sub={a.shape}, diag={b.shape}, "
            f"sup={c.shape}, rhs={d.shape}"
        )
    if not (
        np.all(np.isfinite(a))
        and np.all(np.isfinite(b))
        and np.all(np.isfinite(c))
        and np.all(np.isfinite(d))
    ):
        raise ValueError("thomas_solve: non-finite input")

    # Forward elimination: c'_i and d'_i overwrite copies of c and d.
    cp = np.empty(n, dtype=np.float64)  # modified super-diagonal (cp[n-1] unused)
    dp = np.empty(n, dtype=np.float64)  # modified rhs
    piv = b[0]
    if abs(piv) < _PIVOT_TOL:
        raise ValueError("thomas_solve: zero pivot at row 0")
    cp[0] = c[0] / piv if n > 1 else 0.0
    dp[0] = d[0] / piv
    for i in range(1, n):
        piv = b[i] - a[i - 1] * cp[i - 1]
        if abs(piv) < _PIVOT_TOL:
            raise ValueError(f"thomas_solve: zero pivot at row {i}")
        cp[i] = c[i] / piv if i < n - 1 else 0.0
        dp[i] = (d[i] - a[i - 1] * dp[i - 1]) / piv

    # Back substitution.
    x = np.empty(n, dtype=np.float64)
    x[n - 1] = dp[n - 1]
    for i in range(n - 2, -1, -1):
        x[i] = dp[i] - cp[i] * x[i + 1]
    return x
